### PR TITLE
Telemetry V2 preps (Serialization Refactoring)

### PR DESCRIPTION
**Pull request by ygree**
# What Does This Do

Changes Telemetry's serialization from DTOs to JsonWriter.

# Motivation

It's needed for Telemetry V2 message-batching implementations that would otherwise be impossible because currently one request is represented as a single object (DTO) and JsonWriter will help to construct a request body in more flexible way with the ability to control the size of the request body.

# Additional Notes

Replace TelemetryService tests of using mocks approach to asserting generated actual requests. It will help making it more flexible for upcoming changes and detach tests from internal details.
Also removed bunch of unused classes because they are misleading.

### CODE REVIEW

**Comment by PerfectSlayer** [comment]
Another comment I can't make during the review as it is not one of your changes but, from `TelemetryService` line 208:
```java
while (!state.dependencies.isEmpty()) {
      RequestBuilder rb = requestBuilderProvider.create(RequestType.APP_DEPENDENCIES_LOADED);
      rb.writeHeader();
      rb.writeDependenciesLoadedEvent(state.dependencies.get(maxDepsPerReq));
      rb.writeFooter();
      HttpClient.Result result = httpClient.sendRequest(rb.request());
      if (result == HttpClient.Result.SUCCESS) {
        state.commit();
      } else if (result == HttpClient.Result.NOT_FOUND) {
        state.rollback();
        return;
      } else {
        state.integrations.rollback(); // HERE
        break;
      }
    }
```
I think `rollback()` should be done on `state.dependencies` rather than `state.integrations`.

**Comment by ygree** [comment]
> Another comment I can't make during the review as it is not one of your changes but, from `TelemetryService` line 208:
> 
> ```java
> while (!state.dependencies.isEmpty()) {
>       RequestBuilder rb = requestBuilderProvider.create(RequestType.APP_DEPENDENCIES_LOADED);
>       rb.writeHeader();
>       rb.writeDependenciesLoadedEvent(state.dependencies.get(maxDepsPerReq));
>       rb.writeFooter();
>       HttpClient.Result result = httpClient.sendRequest(rb.request());
>       if (result == HttpClient.Result.SUCCESS) {
>         state.commit();
>       } else if (result == HttpClient.Result.NOT_FOUND) {
>         state.rollback();
>         return;
>       } else {
>         state.integrations.rollback(); // HERE
>         break;
>       }
>     }
> ```
> 
> I think `rollback()` should be done on `state.dependencies` rather than `state.integrations`.
> 
> I wonder if you can't fix the kind of copy/paste issue by refactoring all the while not empty loop with a method like:
> 
> ```java
> <T> void sendMessage(State state, RequestType type, StateList<T> elements, BiConsumer<RequestBuilder, List<T>> messageWriter) {
>     while (!elements.isEmpty()) {
>       RequestBuilder rb = requestBuilderProvider.create(type);
>       rb.writeHeader();
>       messageWriter.accept(rb, elements.get(maxElementsPerReq));
>       rb.writeFooter();
>       HttpClient.Result result = httpClient.sendRequest(rb.request());
> 
>       if (result == HttpClient.Result.SUCCESS) {
>         state.commit();
>       } else if (result == HttpClient.Result.NOT_FOUND) {
>         state.rollback();
>         return;
>       } else {
>         elements.rollback();
>         break;
>       }
>     }
>   }
> ```
> 
> and then call it with all the elements:
> 
> ```java
> sendMessage(state, RequestType.DISTRIBUTIONS, state.distributionSeries, RequestBuilder::writeDistributionsEvent);
> sendMessage(state, RequestType.LOGS, state.logMessages, RequestBuilder::writeLogsEvent);
> ```

You are right! That is a bug in the existing code! Currently, existing tests don't cover this part of functionality at all! I can change it but the message batching implementation that I am working on top of these necessary prep changes will replace this entire method and will address existing issues. This PR is just a refactoring as stated in the description, it's not meant to change existing functionality.

File: telemetry/src/main/java/datadog/telemetry/RequestBuilder.java [hunk 1]
```diff
@@ -1,129 +1,268 @@
...
+      bodyWriter.endObject(); // request
+    } catch (Exception ex) {
+      throw new SerializationException("footer", ex);
     }
   }
+
+  /**
+   * @return associated request. Its body can still be extended with write methods until it's sent.
+   */
+  public Request request() {
```
[the remark this answers is not in the document]
Comment (ygree) [reply telemetry/src/main/java/datadog/telemetry/RequestBuilder.java:254]: I know, that's why I added this comment and deliberately don't use the `build` name. It comes from the fact that the request body is only read when it's read when the request is sent. I'll clarify this in the comment.

File: telemetry/src/test/groovy/datadog/telemetry/HttpClientSpecification.groovy [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  def 'map an http status code to the correct send result'() {
+    when:
+    def result = httpClient.sendRequest(dummyRequest)
+
+    then:
+    result == sendResult
+    1 * okHttpClient.newCall(_) >> mockResponse(httpCode)
+
+    where:
+    httpCode | sendResult
```
Comment (PerfectSlayer) [inline telemetry/src/test/groovy/datadog/telemetry/HttpClientSpecification.groovy:44]: Should you also try generic _parent_ status code like `200` and `400` (as you handle specific `202` and `404`)?
What about `300`?

**Comment by PerfectSlayer** [comment]
> You are right! That is a bug in the existing code! Currently, existing tests don't cover this part of functionality at all! I can change it but the message batching implementation that I am working on top of these necessary prep changes will replace this entire method and will address existing issues. This PR is just a refactoring as stated in the description, it's not meant to change existing functionality.

The refactoring looks good 👍  Some of my comments was mainly to discuss other possible improvements but I understand it is only the first parts of something bigger. Nonetheless, there is still a lot of TODO and comments that needs to be ironed out.

One thing I wonder is should we rather merge this PR into `master` or target a dedicated project branch, like `project/telemetry`? Creating a dedicated project branch will allow to get regular PR reviews approved while leaving the state of telemetry module not fully complete (as it won't impact master). Once we are done the project branch, we can quickly merged it into master. WDYT?

**Comment by ygree** [comment]
> > You are right! That is a bug in the existing code! Currently, existing tests don't cover this part of functionality at all! I can change it but the message batching implementation that I am working on top of these necessary prep changes will replace this entire method and will address existing issues. This PR is just a refactoring as stated in the description, it's not meant to change existing functionality.
> 
> The refactoring looks good 👍 Some of my comments was mainly to discuss other possible improvements but I understand it is only the first parts of something bigger. Nonetheless, there is still a lot of TODO and comments that needs to be ironed out.
> 
> One thing I wonder is should we rather merge this PR into `master` or target a dedicated project branch, like `project/telemetry`? Creating a dedicated project branch will allow to get regular PR reviews approved while leaving the state of telemetry module not fully complete (as it won't impact master). Once we are done the project branch, we can quickly merged it into master. WDYT?

The TODO comments I've added are meant to make flaws in the existing functionality explicit. They are to be addressed in next changes and are not goals for this refactoring.

I'd rather merge this PR into `master` as it's just a refactoring that doesn't add any new functionality.

**Review by PerfectSlayer — APPROVED** [review]
Alright, let's ship it to `master` then 👌  
Feel free to ping me when you push new stuff for telemetry you need review for.

**Pull request merged**